docs/proposals/20220411-cluster-api-state-metrics.md

```diff
@@ -0,0 +1,389 @@
+- `ClusterResourceSetBinding`
+- `MachineHealthCheck`
+- `MachinePool`
+- `KubeadmConfig`
+- `KubeadmConfigTemplate`
+
+The `Cluster` CR will have important information in their status fields similar to [Pod] metrics like `status.Ready` or `status.Conditions`.
+
+Currently it is not important to implement metrics for `KubeadmConfig` and `KubeadmConfigTemplate` because both only contain configuration data (e.g., passed via cloud-init to the machine). However they may be compared to `ConfigMaps` or `Secrets`.
+
+#### How does kube-state-metrics work
+
+Kube-state-metrics exposes metrics by a http endpoint to be consumed by either Prometheus itself or a compatible scraper [[1]].
+
+Since kube-state-metrics v1.5 large performance improvements got introduced to kube-state-metrics which are documented at the [Performance Optimization Proposal](https://github.com/kubernetes/kube-state-metrics/blob/master/docs/design/metrics-store-performance-optimization.md#Proposal). This document also explains the current internals of kube-state-metrics.
+
+It caches the current state of the metrics using an internal cache and updates this internal state on add, update and delete events of watched resources.
+
+On requests to `/metrics` the cached data gets concatenated to a single string and returned as a response.
+
+#### Reuse kube-state-metrics packages
+
+Cluster-api-state-metrics should re-use as many packages provided by kube-state-metrics as possible. This allows re-use of flags, configuration and extended functionality like sharding or tls configuration without additional implementation.
+
+An [extension mechanism](https://github.com/kubernetes/kube-state-metrics/pull/1644) was introduced to the kube-state-metrics packages which allows using its basic mechanism but defining custom metrics for Custom Resources.
+The `k8s.io/kube-state-metrics/v2/pkg/customresource.RegistryFactory`[[2]] interface was [introduced](https://github.com/kubernetes/kube-state-metrics/pull/1644) to allow defining custom metrics for Custom Resources while leveraging kube-state-metrics logic.
+
+Cluster-api-state-metrics will have to implement the `customresource.RegistryFactory` interface for each custom resource.
+The interface defines the function `MetricFamilyGenerators(allowAnnotationsList, allowLabelsList []string) []generator.FamilyGenerator` to be implemented which then contains the specific metric implementations.
+A detailed implementation example is available at the [package documentation](https://pkg.go.dev/k8s.io/kube-state-metrics/v2@v2.4.2/pkg/customresource#RegistryFactory).
+
+These `customresource.RegistryFactory` implementations get used in a `main.go` which configures and starts the application by using `k8s.io/kube-state-metrics/v2/pkg/app.RunKubeStateMetrics(...)`[[3]].
+
+#### Package structure for cluster-api-state-metrics
+
+- `/exp/pkg/store` contains the metric implementation exposed by the `customresource.RegistryFactory`[[2]] interface.
```

Comment:
Is this path supposed to be relative to the project root? If yes, we might have a problem with go module boundaries.

In case we have source code in `/exp/pkg/store` and `/exp/state-metrics` the `go.mod` for these should be in `/exp`, but there is already some code in `/exp` that belong to the root `/go.mod` (`MachinePool`, `ClusterResourceSet`).

I suggest to move `/exp/pkg` to `/exp/state-metrics/pkg`.

Reply:
good finding, no it was supposed to land under `exp/state-metrics/pkg/store` :+1: